src/mqueryfront/src/QueryPage.js

```diff
@@ -24,49 +23,71 @@ class QueryPage extends Component {
     constructor(props) {
         super(props);
 
-        let initialState = { ...INITIAL_STATE };
-        if (this.props.match.params.hash) {
-            initialState.qhash = this.props.match.params.hash;
-        }
-        this.state = initialState;
+        this.state = { ...INITIAL_STATE };
+        this.trackJobTimeout = null;
 
-        this.updateQhash = this.updateQhash.bind(this);
-        this.updateQueryError = this.updateQueryError.bind(this);
-        this.updateQueryPlan = this.updateQueryPlan.bind(this);
         this.collapsePane = this.collapsePane.bind(this);
         this.updateYara = this.updateYara.bind(this);
+        this.setActivePage = this.setActivePage.bind(this);
+        this.submitQuery = this.submitQuery.bind(this);
+        this.editQuery = this.editQuery.bind(this);
+        this.selectTaint = this.selectTaint.bind(this);
+    }
+
+    get queryHash() {
+        return this.props.match.params.hash;
     }
 
     componentDidMount() {
-        if (this.state.qhash) {
-            axios.get(API_URL + "/job/" + this.state.qhash).then((response) => {
-                this.updateQhash(this.state.qhash, response.data.raw_yara);
-            });
+        if (this.queryHash) {
+            this.fetchJob();
         }
         axios.get(API_URL + "/backend/datasets").then((response) => {
             this.setState({ datasets: response.data.datasets });
         });
     }
 
-    componentWillUnmount() {
-        if (this.timeout !== null) {
-            clearTimeout(this.timeout);
-        }
+    async fetchJob() {
+        // Go to the job mode
+        // Load initial job information and start tracking results
+        let response = await axios.get(API_URL + "/job/" + this.queryHash);
+        this.setState(
+            {
+                ...INITIAL_STATE,
+                rawYara: response.data.raw_yara,
+                collapsed: true,
+                job: response.data,
+                datasets: this.state.datasets,
+            },
+            () => {
+                this.trackJob();
+            }
+        );
+    }
 
-        this.setState({
-            qhash: null,
-        });
+    componentWillUnmount() {
+        this.cancelJob();
     }
 
     componentDidUpdate(prevProps, prevState) {
-        if (
-            this.props.match.path === "/" &&
-            this.props.location.key !== prevProps.location.key &&
-            (typeof this.props.location.state === "undefined" ||
-                typeof this.props.location.state.internal === "undefined" ||
-                !this.props.location.state.internal)
-        ) {
-            this.setState(INITIAL_STATE);
+        const prevQueryHash = prevProps.match.params.hash;
+        if (this.queryHash) {
+            if (prevQueryHash !== this.queryHash) {
+                // Went to the job mode or switched to another job
```

Comment:
what is "job mode"?

Reply:
Is it the shift from writing a query to executing a query?

Reply:
Yes. Previously there was `this.state.mode` element that was in two states depending on the phase: `query` (editing) or `job` (executing). Now, it's deduced from the path (`/` for `query` and `/query/<hash>` for `job`). I decided that I'd still refer to that `mode` state in comments to improve code readability.

Reply:
looks good, consider improving the comment maybe. 
nothing too important, i just wondered